lib/min_heap.rb

```diff
@@ -47,25 +53,58 @@ def to_s
   # Time complexity: ?
   # Space complexity: ?
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store[0].nil?
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(logn)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    parent_i = (index - 1) / 2
+    return if parent_i < 0
+    parent = @store[parent_i]
+    current = @store[index]
+
+    while parent_i >= 0 && parent.key > current.key
+      swap(index, parent_i)
+      index = parent_i
+      parent_i = (index - 1) / 2
+      parent = @store[parent_i]
+    end
   end
 
   # This helper method takes an index and 
-  #  moves it up the heap if it's smaller
-  #  than it's parent node.
+  # moves it up the heap if it's smaller
+  # than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    while index < @store.length
+      left_i = index * 2 + 1
+      right_i = index * 2 + 2
+      child_index = nil
+      left = @store[left_i]
+      right = @store[right_i]
+
+      if left && right
+        child_index = left.key < right.key ? left_i : right_i
+        if @store[child_index].key < @store[index].key
+          swap(index, child_index)
+          index = child_index
+        else
+          return
+        end
+      elsif left && !right
```

Comment:
Nice work catching the scenario where the right is not in the array, but left is.